[HUDI-430] Adding InlineFileSystem to support embedding any file format as an InlineFile

nsivabalan opened the pull request:
## *Tips*
- *Thank you very much for contributing to Apache Hudi.*
- *Please review https://hudi.apache.org/contributing.html before opening a pull request.*

## What is the purpose of the pull request
- This PR adds a new FileSystem called InlineFileSystem to support embedding any file format as an InlineFile within a regular file.  InlineFS will be used only in read path. 
- Have added InMemoryFileSystem as part of the PR for write path.

## Verify this pull request

This change added tests and can be verified as follows:

  - *Added tests for InLineFileSystem and InLineFileSystem*
  - *Added tests for testing InlineFS for Parquet and HFile format*

## Committer checklist

 - [ ] Has a corresponding JIRA in PR title & commit
 
 - [ ] Commit message is descriptive of the change
 
 - [ ] CI is green

 - [ ] Necessary doc changes done or have another open PR
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

vinothchandar commented:
Some tests seem to be failing? 

```
Failed tests: 
10649  TestInMemoryFileSystem.testCreateWriteGetFileAsBytes:55 array lengths differed, expected.length=80 actual.length=17967
10650Tests in error: 
10651  TestParquetReadWriteFlow.testSimpleInlineFileSystem:96->readParquetGenericRecords:125 » IO
```

vinothchandar commented:
@nsivabalan is this ready to go?  do we want to close #1068 ?

vinothchandar commented:
@nsivabalan quick update on this? this is ready for review?

vinothchandar commented:
@nsivabalan are you still blocked by the test failures for this one?

nsivabalan commented:
@vinothchandar : Have fixed the build failures. Patch is good to review.

vinothchandar commented:
@nsivabalan awesome. will do..

vinothchandar commented on hudi-utilities/src/main/java/org/apache/hudi/utilities/inline/fs/InMemoryFileSystem.java:
```diff
@@ -0,0 +1,120 @@
+ * A FileSystem which stores all content in memory and returns a byte[] when {@link #getFileAsBytes()} is called
+ * This FileSystem is used only in write path. Does not support any read apis except {@link #getFileAsBytes()}.
+ */
+public class InMemoryFileSystem extends FileSystem {
```
can we move all of this code to `hudi-common`

nsivabalan replied:
Sure. will address along with other feedback. Guess you are not blocked from your review on this.

nsivabalan commented:
@vinothchandar : a reminder to review this patch. Or should I ask someone to pitch in?

vinothchandar commented:
@nsivabalan its on my queue.. while you wait, can you please move the classes to the final locations..

vinothchandar commented on hudi-utilities/src/main/java/org/apache/hudi/utilities/inline/fs/InlineFileSystem.java:
```diff
@@ -0,0 +1,133 @@
+
+  @Override
+  public FSDataInputStream open(Path inlinePath, int bufferSize) throws IOException {
+    Path outerPath = InLineFSUtils.getOuterfilePathFromInlinePath(inlinePath, getScheme());
```
can we introduce a pojo for this? Something that can keep an outerPath, scheme, startOffset and length together.. we can have this utils method return that..

nsivabalan replied:
Within InlineFileSystem, we need to use Path in most places. So even if we introduce a POJO, I feel it may not be of much use. 1: all arguments to this class is Path. So, we have to convert to P
OJO before we do anything. We can't store it as instance variable as well 2: Introducing a POJO might help in fetching start offset and length which is done only once in this class.  Not sure how much of value it adds to define a POJO. But open to making change if you feel so.

vinothchandar commented on hudi-utilities/src/main/java/org/apache/hudi/utilities/inline/fs/InlineFileSystem.java:
```diff
@@ -0,0 +1,133 @@
+    Path outerPath = InLineFSUtils.getOuterfilePathFromInlinePath(inlinePath, getScheme());
+    FileSystem outerFs = outerPath.getFileSystem(conf);
+    FileStatus status = outerFs.getFileStatus(outerPath);
+    FileStatus toReturn = new FileStatus(InLineFSUtils.length(inlinePath), status.isDirectory(), status.getReplication(), status.getBlockSize(),
```
and all of this code will read nicely.. as just getters on that pojo

nsivabalan replied:
have commented above on this.

vinothchandar commented on hudi-utilities/src/test/java/org/apache/hudi/utilities/inline/fs/TestHFileReadWriteFlow.java:
```diff
@@ -0,0 +1,243 @@
+  }
+
+  private FSDataOutputStream createFSOutput(Path name, Configuration conf) throws IOException {
+    //if (fs.exists(name)) fs.delete(name, true);
```
remove?

vinothchandar commented on hudi-utilities/src/test/java/org/apache/hudi/utilities/inline/fs/TestHFileReadWriteFlow.java:
```diff
@@ -0,0 +1,243 @@
+  private FSDataOutputStream createFSOutput(Path name, Configuration conf) throws IOException {
+    //if (fs.exists(name)) fs.delete(name, true);
+    FSDataOutputStream fout = name.getFileSystem(conf).create(name);
+    return fout;
```
single line return?

nsivabalan replied:
Sorry had some debug statements before and hence. will fix it.

vinothchandar commented on hudi-utilities/src/test/java/org/apache/hudi/utilities/inline/fs/TestHFileReadWriteFlow.java:
```diff
@@ -0,0 +1,243 @@
+/**
+ * Tests {@link InlineFileSystem} using HFile writer and reader.
+ */
+public class TestHFileReadWriteFlow {
```
this file has a fair amount of test code.. any way to simply/reuse?

vinothchandar replied:
rename to `TestHFileInlining`

nsivabalan replied:
Most of it is supporting cast for HFile read and write. So, not sure how much we re-use we could do here. Not too much value in modularizing it because the TestParquetInLining may have lot of supporting cast for Parquet.

vinothchandar commented on hudi-utilities/src/test/java/org/apache/hudi/utilities/inline/fs/TestParquetReadWriteFlow.java:
```diff
@@ -0,0 +1,149 @@
+/**
+ * Tests {@link InlineFileSystem} with Parquet writer and reader.
+ */
+public class TestParquetReadWriteFlow {
```
rename `TestParquetInlining`

vinothchandar commented on pom.xml:
```diff
@@ -684,6 +684,11 @@
+      <dependency>
+        <groupId>org.apache.hbase</groupId>
+        <artifactId>hbase-server</artifactId>
+        <version>${hbase.version}</version>
```
scope to be test?

nsivabalan commented on hudi-utilities/src/main/java/org/apache/hudi/utilities/inline/fs/InLineFSUtils.java:
```diff
@@ -0,0 +1,94 @@
+   * @return
+   */
+  public static Path getEmbeddedInLineFilePath(Path outerPath, String origScheme, long inLineStartOffset, long inLineLength) {
+    String subPath = outerPath.toString().substring(outerPath.toString().indexOf(":") + 1);
```
I checked Path, but couldn't find anything. toString() does have scheme if present. 

@Override
  public String toString() {
    // we can't use uri.toString(), which escapes everything, because we want
    // illegal characters unescaped in the string, for glob processing, etc.
    StringBuilder buffer = new StringBuilder();
    if (uri.getScheme() != null) {
      buffer.append(uri.getScheme());
      buffer.append(":");
    }
    if (uri.getAuthority() != null) {
      buffer.append("//");
      buffer.append(uri.getAuthority());
    }
    if (uri.getPath() != null) {
      String path = uri.getPath();
      if (path.indexOf('/')==0 &&
          hasWindowsDrive(path) &&                // has windows drive
          uri.getScheme() == null &&              // but no scheme
          uri.getAuthority() == null)             // or authority
        path = path.substring(1);                 // remove slash before drive
      buffer.append(path);
    }
    if (uri.getFragment() != null) {
      buffer.append("#");
      buffer.append(uri.getFragment());
    }
    return buffer.toString();
  }

nsivabalan commented on hudi-utilities/src/main/java/org/apache/hudi/utilities/inline/fs/InLineFSUtils.java:
```diff
@@ -0,0 +1,94 @@
+ */
+public class InLineFSUtils {
+
+  private static final String INLINE_FILE_STR = "inline_file";
```
Nothing significant. I am just using it to delimit regular path info from inline related attributes.
Eg: "inlinefs://<path_to_outer_file>/s3a/inline_file/?start_offset=20&length=40". If you feel, its not adding much value, I can remove it.

nsivabalan commented on hudi-utilities/src/main/java/org/apache/hudi/utilities/inline/fs/InLineFSUtils.java:
```diff
@@ -0,0 +1,94 @@
+  public static Path getOuterfilePathFromInlinePath(Path inlinePath, String outerScheme) {
+    String scheme = inlinePath.getParent().getParent().getName();
+    Path basePath = inlinePath.getParent().getParent().getParent();
+    return new Path(basePath.toString().replaceFirst(outerScheme, scheme));
```
sorry I don't get you. I haven't used indexOf() in this method at all. not sure I get your feedback.

nsivabalan commented on hudi-utilities/src/main/java/org/apache/hudi/utilities/inline/fs/InLineFSUtils.java:
```diff
@@ -0,0 +1,94 @@
+   * @param outerScheme
+   * @return
+   */
+  public static Path getOuterfilePathFromInlinePath(Path inlinePath, String outerScheme) {
```
actually you are right. will fix it.

nsivabalan commented on hudi-utilities/src/main/java/org/apache/hudi/utilities/inline/fs/InlineFileSystem.java:
```diff
@@ -0,0 +1,133 @@
+ * - Inlined path is of the form "inlinefs:///path/to/outer/file/<outer_file_scheme>/inline_file/?start_offset=<start_offset>&length=<length>
+ * <p>
+ * TODO: The reader/writer may try to use relative paths based on the inlinepath and it may not work. Need to handle
+ * this gracefully eg. the parquet summary metadata reading. TODO: If this shows promise, also support directly writing
```
actually I wanted to sync up with you on this. Will sycn up offline.

nsivabalan commented on hudi-utilities/src/main/java/org/apache/hudi/utilities/inline/fs/InlineFileSystem.java:
```diff
@@ -0,0 +1,133 @@
+ */
+public class InlineFileSystem extends FileSystem {
+
+  static final String SCHEME = "inlinefs";
```
yes, I have fixed the scheme issue. Can you check once I have addressed all comments.

nsivabalan commented:
@vinothchandar : have addressed your comments for the most part. I will wait for 2 days for you to have a final look. If not, will merge it myself as per your suggestion.

vinothchandar commented:
@nsivabalan Only one issue I see

I would advise against use of special string in parsing.. if someone for e,g has `inline_file` in the `<path_to_outer_file>`, then we are prone to errors? Can you fix it with along these lines?

then once CI parses, please go ahead merge
